### ISSUE TITLE
Reverse differentiable arccos

### ISSUE THREAD

**Issue by unalmis** [google/jax#23649]
### Description

Using JAX's nan debugger, I have learned that an nan leaked from the following code
```python
x = jnp.arcos(jnp.clip(y, -1, 1))
```
The solution was
```python
eps = jnp.finfo(jnp.array(1.0).dtype).eps  # or similar
x = jnp.arccos(jnp.clip(y, -1 + eps, 1 - eps))
```

I am reporting this because the first should be sufficient to avoid nan, perhaps there is an easy fix in the implementation of arccos?

### System info (python version, jaxlib version, accelerator, etc.)

```
An NVIDIA GPU may be present on this machine, but a CUDA-enabled jaxlib is not installed. Falling back to cpu.
jax:    0.4.30
jaxlib: 0.4.28
numpy:  1.26.4
python: 3.10.14 | packaged by conda-forge | (main, Mar 20 2024, 12:45:18) [GCC 12.3.0]
jax.devices (1 total, 1 local): [CpuDevice(id=0)]
process_count: 1
platform: uname_result(system='Linux', node='fedora', release='6.10.9-200.fc40.x86_64', version='#1 SMP PREEMPT_DYNAMIC Sun Sep  8 17:23:55 UTC 2024', machine='x86_64')
```

**Comment by imSanko** [comment]
How Can I contribute to fix this issue or if don't mind can you guide me through @unalmis

**Comment by jakevdp** [comment]
Thanks for the question! I think the gradient here is working correctly. I'm assuming you're seeing something like this:
```python
>>> def f(x):
...   return jnp.arccos(jnp.clip(x, -1, 1))
>>> jax.grad(f)(10.0)
nan
```
Symbolically, you're writing

$f(x) = g(h(x))$

where *g* is `arccos` and *h* is `clip`. By the chain rule, the gradient is

$f\prime(x) = h\prime(x) g\prime(h(x))$

Where, in the case above, $x=10$, so $h(x)=1$. The derivative terms are $h\prime(x) = 0$, and $g\prime(1) = -\inf$. The result then is $0 * -\inf$ which results in `NaN`.

The fix for this will depend on the context. In some cases, you may instead wish to clip to the range `[-0.99, 0.99]` so the arccos gradient never hits infinity. In others, you may want to use a smooth alternative to the clip function, such as a sigmoid modified to the range [-1, 1].

What do you think?

**Comment by unalmis** [comment]
I see thanks. The context for this was https://jax.readthedocs.io/en/latest/faq.html#gradients-contain-nan-where-using-where; I see now strict inequalities are used there so there is no inconsistency.

**Comment by unalmis** [comment]
Is it possible for the gradient to become nan from floating point errors?
I'm wondering if the finite sized buffer offered by
```python
eps = jnp.finfo(jnp.array(1.0).dtype).eps
x = jnp.arccos(jnp.clip(y, -1 + eps, 1 - eps))
```
is ever safer than the differential buffer offered by
```python
x = jnp.arccos(jnp.where(jnp.abs(y) < 1, y, 0))
```
in the context of the double where trick?

**Comment by jakevdp** [comment]
> Is it possible for the gradient to become nan from floating point errors?

Yes, for example if your gradient involves multiplying (small number) * (big number) and due to floating point precision, the small number goes to zero and the big number goes to infinity. Is that what you're asking about?

Regarding the two approaches you mentioned: they return different results for inputs $|y| \ge 1$, but both seem like reasonable strategies for avoiding the NaN gradients you encountered above.
